### ISSUE TITLE
How to deal with POMDP problem?

### ISSUE THREAD

**Issue by gzrjzcx** [Unity-Technologies/ml-agents#2238]
Hi,
How to solve a POMDP problem? 
As doc said, only PPO algorithm is implemented here, if I want to find a POMDP solution, should I implement the relevant algorithm using the Python API? Or any method to solve POMDP environment?
  
For example, in terms of the classical `Tiger` example for POMDP, 
![POMDP_TIger](https://user-images.githubusercontent.com/33338567/60923303-8f087100-a296-11e9-8952-c46abcaab698.jpg)
Do I need implement the value iteration by myself?

**Comment by unityjeffrey** [comment]
CC: @ervteng

**Comment by ervteng** [comment]
Hi @gzrjzcx, LSTM is a classic way of dealing with POMDPs in deep RL (see: https://arxiv.org/abs/1507.06527). You can turn this on using `use_recurrent` and setting a non-zero sequence length. You can also use vector stacking to ameliorate this issue. 

Many problems that may appear to be partially observable (e.g. our Pyramids environment) are still solvable with vanilla PPO, however, so give that a go as well before moving on to the more advanced approaches.

**Comment by roboserg** [comment]
Shouldn't vanilla PPO solve this tiger example w/o problems, since no memory is needed and the location of t he tiger is given by a direct observation?

**Comment by gzrjzcx** [comment]
> Hi @gzrjzcx, LSTM is a classic way of dealing with POMDPs in deep RL (see: https://arxiv.org/abs/1507.06527). You can turn this on using `use_recurrent` and setting a non-zero sequence length. You can also use vector stacking to ameliorate this issue.
> 
> Many problems that may appear to be partially observable (e.g. our Pyramids environment) are still solvable with vanilla PPO, however, so give that a go as well before moving on to the more advanced approaches.

Thanks @ervteng, I will try PPO firstly.
BTW what the meaning of the 'vector stacking' to ameliorate the issue? In the Pyramid example, the codes
```
            const float rayDistance = 35f;
            float[] rayAngles = {20f, 90f, 160f, 45f, 135f, 70f, 110f};
            float[] rayAngles1 = {25f, 95f, 165f, 50f, 140f, 75f, 115f};
            float[] rayAngles2 = {15f, 85f, 155f, 40f, 130f, 65f, 105f};

            string[] detectableObjects = {"block", "wall", "goal", "switchOff", "switchOn", "stone"};
            AddVectorObs(rayPer.Perceive(rayDistance, rayAngles, detectableObjects, 0f, 0f));
            AddVectorObs(rayPer.Perceive(rayDistance, rayAngles1, detectableObjects, 0f, 5f));
            AddVectorObs(rayPer.Perceive(rayDistance, rayAngles2, detectableObjects, 0f, 10f));
```
why need to use Perceive to make rayDistance, rayAngles etc. as  a list?  

Cheers

**Comment by roboserg** [comment]
the list just defines the angles to shoot lasers from.

**Comment by gzrjzcx** [comment]
Hi @roboserg, thanks for your explanation. That is to say, if I want to use LSTM, I can add position(t) position(t-1) position(t-2) etc as the observations to  ameliorate the issue?

**Comment by roboserg** [comment]
With LSTM you dont need stacking vectors, since the LSTM has a memory build in (hidden state). You would use stacking observations only if you _dont_ use an LSTM but still want some form of (short) memory. LSTM is a better choice, if you need memories over a long time horizon.

**Comment by gzrjzcx** [comment]
Gotcha, thanks. 
In addition, how do u set the state, action space with PPO in terms of the Tiger problem? In other words, how do you deal with the observations(belief state, i.e. the listening results here)?

**Comment by roboserg** [comment]
You would `AddVectorObs(observation)` where observation is either 0 or 1. And you would have two discreet actions "pick left door" and "pick right door"

**Comment by gzrjzcx** [comment]
> You would `AddVectorObs(observation)` where observation is either 0 or 1. And you would have two discreet actions "pick left door" and "pick right door".
> 
> You would need to recreate the tiger environment in Unity so that the agent can interact with it.

But for POMDP, you don't know which the specific state you are, therefore you need to use 'listen' action to observe the environment. That is to say there are three actions you can take, and the 'listen' action will result two observations(tiger-left, tiger-right) which are possible to influence the further action choosing. Is your method adapt to this?

**Comment by roboserg** [comment]
well yes, so you would have observation of 0 if your agents doesnt know the state its in, and say -1 and +1 for tiger left and right. And as you said, 3 actions.

**Comment by ervteng** [comment]
If I understand correctly, the Tiger problem is a POMDP because your "observation" (i.e. whether you hear the tiger on the left or on the right) only corresponds to a _probability_ that the true state is one way or another, but you don't know for sure. So the right strategy would be to listen multiple times and use the history of actions to determine which door to open, as each listen will increase the probability that you're right about what you believe the state to be. Vector stacking or LSTM will help with this. Note that vector stacking is turned on in the LearningBrain in Unity (called Stacked Vectors) and LSTM is turned on in the `trainer_config.yaml` on the Python side. 

@gzrjzcx, you're right, 3 actions - open door (left or right), listen. And 2 observations - (hear tiger left, hear tiger right). When open door is selected you call Done() and end the episode, so it won't matter what the following observation is.

**Comment by gzrjzcx** [comment]
@ervteng Thanks!
That is why I am confused. That is to say, in POMDP context, the `observations` are the results of listen action( plus the probability distribution form belief states right? ), but ML-Agents also deals with observations as MDP right?  
  
In term of the Tiger problem, we only need to add the listen results as the observations, and give a light cost for listen action, then no need to consider how listen results influence the action selection?

**Comment by ervteng** [comment]
Right, in ML-Agents (and most RL) observations are the results of the action. Hence, when you take an action, you get an observation. Then, the algorithm needs decide what the next action is. In your problem, only the listen action gets an observation. However, the other two actions are terminal (i.e. they end the episode), so you can assign any observation to them - this observation will never be used because there is no next state.

**Comment by gzrjzcx** [comment]
Hi @ervteng ,  what the meaning of ' assign any observation to them' ? It means that these two terminal actions can result any observation?
  
The other problem is that what is the difference between stacked vector and LSTM? 
  
Stacked vectors are the number of previous vector observations for decision making, it means that it only stacked vector observations and give them to python API?  As for the Hallway example, the stacked vector is set to 3, which means that when making decision at timestep t,  we use Observation(t), Observation(t-1) and Observation(t-2) collectively? How you can know that 3 stacked vector is enough for agent to remember the target color?
  
In terms of the LSTM, it is that when training, using the Recurrent Neural Network(replacing the fully connected layer as a current LSTM layer ) so that model can use a `sequence_length` experiments for learning instead of single experiment? Is `sequence_length` parameters need to be equal to the stacked vectors? (E.g., should set to 3 in Hallway examples)?  
And is the Hallway example employed LSTM techniques? How to set the `sequence_length` and `memory_size` parameters so that the memory is enough to remember the target color and go to the matched gate? I mean, the memory of target color will not disappear until it reached the correct color gate.

**Comment by gzrjzcx** [comment]
HI @ervteng,
According to the Hallway example, I have trained a model for Tiger problem successfully. And it works as expected. 
But now there still are two problems:  
1. Now, each episode is too long. It means that the agent needs to listen so many times then it can make a decision. I want  to limit the number of listen action in 5 times. Therefore,  I try to set the agent `max_step` to 5, is it correct? Because I think that is the `max_step` is set to 5, then each episode is no more than 5 steps. 
2. How to set LSTM and Stacked vectors parameters.
I have checked the Hallway example trainer config, the LSTM parameter is using the default set, i.e. `sequence_length = 64` and `memory_size = 256`. Then as for the 'stacked vectors', I set it to 5(because the `max_step` is 5, I want to make a decision within 5 times). However, the problem now is the mean reward is not increasing. (Or increasing very slow...). The first success model only need 1m steps then the mean reward started to increase fast. However, for this model even 3m steps the mean reward still slightly decreased. 
  
I know the `max_step=5` is not correct now and makes it is so difficult to learn something. But how could I limit the number of listen actions?  
By the way, the Hallway example also add the `current_step / max_step` as observation, I am confused why need this? And what is the impact of this? In addition, the `stacked vectors = 3` for Hallway example, is that means that the once the  three observation(t),  observation(t-1), observation(t-2) contains the detected target color(e.g. orange target), then the agent know go to orange gate to get reward?
  
Cheers

**Comment by ervteng** [comment]
Hi @gzrjzcx, shouldn't the agent be incentivized to listen less times, since it gets a negative reward each time? If it is listening a lot, it might have learned that it is far less risky to listen a lot than to risk running into the tiger. You can play with the size of the negative reward to see if there's a difference.

I would also cap out the reward to +/-1 to make the learning more stable. You can just divide all of the rewards by 100 to do this. You are correct that setting max_step to 5 will force the episode to be at most 5 steps long, but I'd set it longer and let the agent learn to shorten the episode itself.

stacked vectors equal to the max_step might be problematic. You probably also don't need both stacked vectors and LSTM, one or the other should be OK. 

Current_step/max_step gives the agent a notion of time. I don't think you'd need this in your problem; in fact, I believe Hallway is solvable without it.

**Comment by gzrjzcx** [comment]
Hi @ervteng,
Thanks for your reply.
Yeah, in fact the agent has learned to listen less times, because I set the `max_step=1000`, and even longer episodes have only about 15 listen actions. But it is still too long for my target, I want to let it learn to make decision within 5 steps. So, any clue to tune it? I will try to add the negative reward for each listen action.
  
And today I have also trained a model with same settings with previous successful one, but used the LSTM and stacked vectors(set to 5). But the result is not good, the model is not stable, the result can be shown below:
<img width="696" alt="屏幕快照 2019-07-17 上午2 13 45" src="https://user-images.githubusercontent.com/33338567/61340315-ae287500-a839-11e9-9893-b094f80ff794.png">
It looks very unstable, like there are two lines but it is just very acutely fluctuant. It means that the first 1000 steps the mean reward is 0.7, but the next 1000 steps the mean reward will reduce to -0.4 or lower. I have no idea that why only adding the LSTM and stacked vectors techniques, will lead to this problem. But the model looks have learned something certainly. 
  
I really use the +/-1 scale reward, a small problem is that for the +/-1 reward, is the  mean reward possible to more than 1? 
  
I am working on the POMDP problem that I need to use previous 3 or 5 observations to make a decision. That is to say, the next decision is influenced by the previous 3 observations, is the stacked vector enough for this case?
